tidy3d/components/boundary.py

```diff
@@ -0,0 +1,626 @@
+"""Defines electromagnetic boundary conditions"""
+from abc import ABC
+from typing import Union, Tuple, List
+
+import pydantic as pd
+import numpy as np
+
+from .base import Tidy3dBaseModel
+from .types import Complex, Axis, Literal
+from .source import GaussianBeam, ModeSource, PlaneWave
+from .medium import Medium
+
+from ..log import SetupError
+from ..constants import EPSILON_0, MU_0, PML_SIGMA
+
+
+class BoundaryEdge(ABC, Tidy3dBaseModel):
+    """Electromagnetic boundary condition at a domain edge."""
+
+    name: str = pd.Field(None, title="Name", description="Optional unique name for boundary.")
+
+
+# PBC keyword
+class Periodic(BoundaryEdge):
+    """Periodic boundary condition class."""
+
+
+# PEC keyword
+class PECBoundary(BoundaryEdge):
+    """Perfect electric conductor boundary condition class."""
+
+
+# PMC keyword
+class PMCBoundary(BoundaryEdge):
+    """Perfect magnetic conductor boundary condition class."""
+
+
+# """ Bloch boundary """
+
+# sources from which Bloch boundary conditions can be defined
+BlochSourceType = Union[GaussianBeam, ModeSource, PlaneWave]
+
+
+class BlochBoundary(BoundaryEdge):
+    """Specifies a Bloch boundary condition along a single dimension.
+
+    Example
+    -------
+    >>> bloch = BlochBoundary(bloch_vec=1)
+    """
+
+    bloch_vec: Complex = pd.Field(
+        ...,
+        title="Normalized Bloch vector component",
+        description="Normalized component of the Bloch vector in units of "
+        "2 * pi / (size along dimension) in the background medium, "
+        "along the dimension in which the boundary is specified.",
+    )
+
+    @property
+    def bloch_phase(self) -> Complex:
+        """Returns the forward phase factor associated with `bloch_vec`."""
+        return np.exp(1j * 2.0 * np.pi * self.bloch_vec)
+
+    @classmethod
+    def from_source(
+        cls, source: BlochSourceType, domain_size: float, axis: Axis, medium: Medium = None
+    ):
```

Comment:
add return type

Reply:
This gives a pytlint error when I add `-> BlochBoundary`, possibly because the class definition is not yet complete?

Reply:
yea, best to just put it in quotes in this situation. it's a known thing that pylint is annoying about return type being type of self or cls